crates/pallet-domains/src/staking_epoch.rs

```diff
@@ -664,93 +332,117 @@ pub(crate) fn do_finalize_slashed_operators<T: Config>(
                 .current_total_stake
                 .checked_add(&operator.current_epoch_rewards)
                 .ok_or(TransitionError::BalanceOverflow)?;
+            let total_shares = operator.current_total_shares;
+            let share_price = Perbill::from_rational(total_shares, total_stake.into());
 
             // transfer all the staked funds to the treasury account
             // any gains will be minted to treasury account
-            Nominators::<T>::drain_prefix(operator_id).try_for_each(|(nominator_id, _)| {
-                let locked_amount = T::Currency::balance_on_hold(&staked_hold_id, &nominator_id);
-                T::Currency::transfer_on_hold(
-                    &staked_hold_id,
-                    &nominator_id,
-                    &T::TreasuryAccount::get(),
-                    locked_amount,
-                    Precision::Exact,
-                    Restriction::Free,
-                    Fortitude::Force,
-                )
-                .map_err(|_| TransitionError::RemoveLock)?;
-
-                total_stake = total_stake
-                    .checked_sub(&locked_amount)
-                    .ok_or(TransitionError::BalanceUnderflow)?;
+            Deposits::<T>::drain_prefix(operator_id).try_for_each(
+                |(nominator_id, mut deposit)| {
+                    let locked_amount =
+                        T::Currency::balance_on_hold(&staked_hold_id, &nominator_id);
 
-                Ok(())
-            })?;
+                    // convert any previous epoch deposits
+                    do_convert_previous_epoch_deposits::<T>(
+                        operator_id,
+                        &mut deposit,
+                        domain_epoch,
+                    )?;
+
+                    // there maybe some withdrawals that are initiated in this epoch where operator was slashed
+                    // then collect and include them to find the final stake amount
+                    let (amount_ready_to_withdraw, shares_withdrew_in_current_epoch) =
+                        calculate_withdraw_share_ssc::<T>(operator_id, nominator_id.clone());
```

Comment:
Related to #1928, if a nominntor has withdrawn all (or partial) of its stakes and the unlocking period is past but forgets to call `unlock_funds` then his stake will still be slashed, although this way the nominntor is incentive to unlock its fund as soon as the unlocking period is over but it is still a bad UX IMO image a nominntor withdrawn all its fund years ago but may still get slash.

Reply:
Okay, take another look it won't lead to the #1928 issues if the nominator withdraws all of its stakes and exit from the operator pool since the `Deposits` is removed for such nominator so we can't find the nominator and call `calculate_withdraw_share_ssc` for it.

But it seems led to another issue, because `Deposits` is removed for fully withdraw nominator, we can't slash them even if the withdrawal is still in the unlocking period.